feat: support multiline for EditableText

gabriel-amram opened the pull request:
This PR introduces the ability to support multiline in EditableText
if multiline is passed it will render a "textarea" instead of an input that cannot be resized by the user, and instead will grow/shrink according to its content up to a certain max-width or max-height defined by the container.

When using Shift+Enter a new line will be added. Hitting Enter will defocus and change the value

This is useful for users who might want to have a long description and would like to be able to edit it and change the content with line breaks.

When multiline is used, ellipsis is removed from the "view" block and from the "textarea".

- [X] I have read the [Contribution Guide](../CONTRIBUTING.md) for this project.

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -201,8 +273,16 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
         >
-          {inputValue || placeholder}
+          {(multiline
+            ? inputValue.split("\n").map((line, index) => (
+                <React.Fragment key={index}>
+                  {line}
+                  {index < inputValue.split("\n").length - 1 && <br />}
+                </React.Fragment>
+              ))
```
Why do we need this? Textareas know how to break lines

gabriel-amram replied:
this is not a textarea, that is the TypographyComponent (Text) that is shown once the textarea is not in edit mode any more

talkor replied:
why not just set ellipsis to false?

gabriel-amram replied:
ellipsis is for wrapping
we're trying to have "new lines" supported
a div needs <br/> to break into a new line

gabriel-amram replied:
by the way it is with ellipsis false

talkor replied:
You can use only `white-space: pre-wrap` and it will behave the same as in editing mode and avoid splitting the string with `<br />`s

gabriel-amram replied:
@talkor works like a charm
changed it

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+
+    /* Dynamically resizes the textarea to fit its content */
+    function resizeTextarea() {
+      console.log("*************************resize called, why?", multiline, inputRef.current);
```
Please remove :)

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -188,7 +245,22 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+        aria-label={ariaLabel}
+        data-testid={dataTestId}
+        className={cx(styles.editableTypography, className)}
+        role={isEditing ? null : "button"}
```
It will not be shown when not editing, plus role null is bad for a11y

gabriel-amram replied:
that part existed already, didn't want to change as i don't know what it breaks, but lets discuss and i'll change accordingly

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+            return;
+          }
+
+          const computedStyle = window.getComputedStyle(textarea);
```
`getComputedStyle` is pretty rough for performance, and this will be called on each render

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+      console.log("*************************resize called, why?", multiline, inputRef.current);
+      if (inputRef.current) {
+        // Temporarily set the height to "auto" to accurately measure the scroll height of the content inside the textarea.
+        setInputHeight("auto");
```
This might cause layout thrashing

gabriel-amram replied:
i am requesting an animation frame to decrease the chances of this to happen

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+    }
+
+    /* Dynamically resizes the textarea to fit its content */
+    function resizeTextarea() {
```
We already do something similar with the width, will it work for the height as well?

gabriel-amram replied:
the width is dependent on the "div" behind the scenes that is growing in width, it is not growing in height as much as i wanted it to :)

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+          const computedStyle = window.getComputedStyle(textarea);
+
+          // Calculate the appropriate height by taking into account the scrollable content inside the textarea,
+          // as well as the styles applied to it, such as padding and border widths.
+          const lineHeight = parseFloat(computedStyle.lineHeight) || 16;
+          const paddingTop = parseFloat(computedStyle.paddingTop) || 0;
+          const paddingBottom = parseFloat(computedStyle.paddingBottom) || 0;
+          const borderTopWidth = parseFloat(computedStyle.borderTopWidth) || 0;
+          const borderBottomWidth = parseFloat(computedStyle.borderBottomWidth) || 0;
+
+          const newHeight = textarea.scrollHeight + borderTopWidth + borderBottomWidth;
+          const minHeight = lineHeight + paddingTop + paddingBottom + borderTopWidth + borderBottomWidth;
```
This feels really fragile, the line height can change and it will break the logic it, the calculation relies on spacings and borders which can also change
There are better stable alternative ways to adjust the height :)

gabriel-amram replied:
couldn't find any, but lets discuss

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.module.scss:
```diff
@@ -26,6 +26,12 @@
   }
 
+  .textarea {
+    max-height: inherit;
```
Where is it defined?

gabriel-amram replied:
it is to allow the user to define a max-height that we can respect
if there is none, it will let it grow as much as needed

talkor replied:
If it's a feature it should be via a prop, not by className

gabriel-amram replied:
you're right, or instead user can commandeer the container and limit its size

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.module.scss:
```diff
@@ -26,6 +26,12 @@
+  .textarea {
+    max-height: inherit;
+    resize: none;
+    box-sizing: border-box;
```
We already use border-box for all elements

talkor commented on packages/core/src/components/EditableText/__stories__/EditableText.stories.tsx:
```diff
@@ -102,3 +102,25 @@ export const Types = {
   }
 };
+
+const multilineValue = "This is a multiline\nhere's the second line";
```
This will not be shown in the story code, let's put it inline

talkor commented on packages/core/src/components/EditableText/__stories__/EditableText.stories.tsx:
```diff
@@ -102,3 +102,25 @@ export const Types = {
+        weight={EditableText.weights.NORMAL}
+        multiline
+        value={multilineValue}
+        className={styles.editableText}
```
Let's avoid using css in stories

gabriel-amram replied:
i am using here the same CSS class other stories here are using

talkor commented on packages/core/src/components/EditableText/__tests__/EditableText.test.tsx:
```diff
@@ -18,6 +18,17 @@ describe("EditableText", () => {
+
+    const input = screen.queryByRole("input");
+    expect(input).toBeInTheDocument();
+    expect(input.tagName).toBe("TEXTAREA");
```
Feels redundant IMO, better just to check that the role is `textbox`  (not `input`)

talkor reviewed (changes requested):
Thank you for the contribution! Great addition
Please see my comments :)

gabriel-amram commented on packages/core/src/components/EditableText/__stories__/EditableText.stories.module.scss:
```diff
@@ -14,3 +14,13 @@
+  flex-direction: row;
+
+  .editableText {
+    width: 400px;
+    max-height: 200px;
```
just to show the max-height capabilities that it can stop growing, i can remove this if its not necessary for the story...

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -43,6 +43,8 @@ export interface EditableTypographyProps extends VibeComponentProps, EditableTyp
   /** Sets the Text/Heading weight */
   weight?: TextWeight | HeadingWeight;
+  /** Controls whether a textarea or a simple input would be rendered, allowing multi-lines */
+  multiline?: boolean;
```
understand, but how would we use this otherwise in the Typography that both EditableText and EditableHeading are using?

talkor replied:
Nevermind my mistake, it won't be shown in the EditableHeading so all good

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +146,56 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
         inputRef.current?.focus();
+
+        if (multiline) {
+          resizeTextarea();
```
it means the textarea just gained focus and we need to calculate its height

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -160,22 +209,30 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
-        onKeyDown={toggleKeyboardEditMode}
-      >
-        {isEditing && (
+    function getEditableElement() {
```
yes

talkor replied:
This is bad for performance

gabriel-amram replied:
i think i found another way, i'll test it and we'll talk about it in our meeting

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -160,22 +209,30 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+            aria-label={ariaLabel}
+            placeholder={placeholder}
+            style={{ width: inputWidth, height: inputHeight }}
+            role="input"
```
changed

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -188,7 +245,22 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+    }
+
+    return (
+      <div
```
this div existed, it is not new, so shouldn't break anything

talkor replied:
Sorry my bad, the diff was confusing :)

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -160,22 +209,30 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+            placeholder={placeholder}
+            style={{ width: inputWidth, height: inputHeight }}
+            role="input"
+            rows={1}
```
no, its ok and lines break. what this does is make the textarea start like a normal input (1 line), as browser defaults is 2 lines

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +148,47 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+          setInputHeight(
+            Math.max(
+              textarea.scrollHeight + textareaBorderBoxSizing.current,
+              textareaLineHeight.current + textareaBorderBoxSizing.current
```
@talkor we'll talk about it in our session, but i refactored this a bit - 
only when focusing i am calculating the getComputedStyle and "cache" it in two refs that i need to use
that way i don't need to call if on every render

gabriel-amram commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -138,15 +148,47 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+        // Temporarily set the height to "auto" to accurately measure the scroll height of the content inside the textarea.
+        setInputHeight("auto");
+
+        requestAnimationFrame(() => {
```
check layoutEffect instead?

gabriel-amram replied:
layout effect does the same thing, and even adds unnecessary calculations on the lineHeight
dropped it
@talkor

talkor commented on packages/core/src/components/EditableTypography/EditableTypography.tsx:
```diff
@@ -141,14 +151,46 @@ const EditableTypography: VibeComponent<EditableTypographyProps, HTMLElement> =
+          }
+
+          // Ensure we at least have 1 line
+          setInputHeight(
+            Math.max(
+              textarea.scrollHeight + textareaBorderBoxSizing.current,
+              textareaLineHeight.current + textareaBorderBoxSizing.current
```
Just one thought, looks like the only difference in the dimensions is 2px, and there's no difference in paddings/line heights etc between the different variants, so looks like doing just:
`setInputHeight(textarea.scrollHeight + 2);` works exactly the same and it can save the getComputeStyle and the other calculations in `calculateTextareaHeightAttrs` (which already have some default static numbers). Giving it `auto` in the beginning will already make it take the the currect box size. WDYT?

gabriel-amram replied:
actually the question gets back to you :)
technically its only 2px because the border-top/bottom-width is 1px (or 2px overall)
border-box should have included the border as well as the padding, but it doesn't

do we want to just rely on always having this border/padding setting and that the user cannot add anything on their own using overwriting css?
@talkor

talkor replied:
Well it is border-box, it's just that `scrollHeight` is based purely on the content and padding of the element's scrollable content, without the borders. We don't want users to override CSS as it can make design inconsistent so I wouldn't take it into consideration, generally speaking we support the variants we officially provide so I think it's ok :)

talkor commented:
Looks great @gabriel-amram ! Only one thought which I commented, and I did some quick QA and found two things worth checking:
- When focused on the textarea and clicking outside in the same block it will not blur
- When going into edit mode the cursor will be in the beginning of the text instead of the last character which I think makes the experience better

talkor commented:
> * When focused on the textarea and clicking outside in the same block it will not blur

<img width="503" alt="Screenshot 2024-12-31 at 11 49 50" src="https://github.com/user-attachments/assets/8ec40ffa-b989-48c8-899a-190bbb34b7be" />
I mean when clicking on the purple area. I think it's because the 400px width, probably changing it to `max-width` instead will fix it

gabriel-amram commented:
> Looks great @gabriel-amram ! Only one thought which I commented, and I did some quick QA and found two things worth checking:
> 
> * When focused on the textarea and clicking outside in the same block it will not blur
> * When going into edit mode the cursor will be in the beginning of the text instead of the last character which I think makes the experience better

@talkor 
the first one, is just a story thing, i removed the wrapping div so its clearer and it works
the second one, this is the default behaviour of browsers (input on focus goes to end of text, textarea to start). i added a piece of code on focus that makes sure that we always go to end of text for both
(this might feel broken though if people are more used to one rather than the other)

talkor reviewed (approved):
Thank you!
IMO worth simplifying the height calculation nevertheless
